File: modules/qr-ec-analyzer/ios/QrEcAnalyzerModule.swift
import ExpoModulesCore
import ZXingObjC
import UIKit

public class QrEcAnalyzerModule: Module {
  public func definition() -> ModuleDefinition {
    Name("QrEcAnalyzer")

    AsyncFunction("analyzeQrCode") { (imagePath: String, promise: Promise) in
      self.analyzeQrCodeInternal(imagePath: imagePath, promise: promise)
    }
  }

  private func analyzeQrCodeInternal(imagePath: String, promise: Promise) {
    DispatchQueue.global(qos: .userInitiated).async {
      do {
        // Load image from path
        guard let image = UIImage(contentsOfFile: imagePath) else {
          promise.resolve([
            "success": false,
            "error": "Failed to load image from path: \(imagePath)"
          ])
          return
        }

        guard let cgImage = image.cgImage else {
          promise.resolve([
            "success": false,
            "error": "Failed to get CGImage"
          ])
          return
        }

        // Create ZXing luminance source
        let source = ZXCGImageLuminanceSource(cgImage: cgImage)
        let binarizer = ZXHybridBinarizer(source: source)
        let bitmap = ZXBinaryBitmap(binarizer: binarizer)

        // Configure hints
        let hints = ZXDecodeHints()
        hints.tryHarder = true

        // Try to decode
        let reader = ZXQRCodeReader()

        var result: ZXResult?
        var error: NSError?

        result = try? reader.decode(bitmap, hints: hints)

        if result == nil {
          // Try with inverted image
          if let invertedSource = source?.invert() {
            let invertedBinarizer = ZXHybridBinarizer(source: invertedSource)
            let invertedBitmap = ZXBinaryBitmap(binarizer: invertedBinarizer)
            result = try? reader.decode(invertedBitmap, hints: hints)
          }
        }

        if let result = result {
          var ecLevel: String? = nil

          // Extract EC level from result metadata
          if let metadata = result.resultMetadata {
            if let ecLevelValue = metadata[NSNumber(value: ZXResultMetadataType.errorCorrectionLevel.rawValue)] {
              ecLevel = String(describing: ecLevelValue)
            }
          }

          promise.resolve([
            "success": true,
            "data": result.text ?? "",
            "ecLevel": ecLevel as Any
          ])
        } else {
          promise.resolve([
            "success": false,
            "error": "No QR code found"
          ])
        }
      } catch {
        promise.resolve([
          "success": false,
          "error": error.localizedDescription
        ])
      }
    }
  }
}

File: modules/qr-ec-analyzer/android/src/main/java/expo/modules/qrecanalyzer/QrEcAnalyzerModule.kt
package expo.modules.qrecanalyzer

import android.graphics.BitmapFactory
import android.net.Uri
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import com.google.zxing.*
import com.google.zxing.common.HybridBinarizer
import com.google.zxing.qrcode.QRCodeReader
import com.google.zxing.qrcode.decoder.ErrorCorrectionLevel
import java.io.File
import java.io.FileInputStream

class QrEcAnalyzerModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("QrEcAnalyzer")

    AsyncFunction("analyzeQrCode") { imagePath: String, promise: Promise ->
      try {
        val result = analyzeQrCodeInternal(imagePath)
        promise.resolve(result)
      } catch (e: Exception) {
        promise.resolve(mapOf(
          "success" to false,
          "error" to (e.message ?: "Unknown error")
        ))
      }
    }
  }

  private fun analyzeQrCodeInternal(imagePath: String): Map<String, Any?> {
    // Load the image
    val file = File(imagePath)
    if (!file.exists()) {
      return mapOf(
        "success" to false,
        "error" to "File not found: $imagePath"
      )
    }

    val inputStream = FileInputStream(file)
    val bitmap = BitmapFactory.decodeStream(inputStream)
    inputStream.close()

    if (bitmap == null) {
      return mapOf(
        "success" to false,
        "error" to "Failed to decode image"
      )
    }

    // Convert bitmap to ZXing format
    val width = bitmap.width
    val height = bitmap.height
    val pixels = IntArray(width * height)
    bitmap.getPixels(pixels, 0, width, 0, 0, width, height)

    val source = RGBLuminanceSource(width, height, pixels)
    val binaryBitmap = BinaryBitmap(HybridBinarizer(source))

    // Configure hints for better decoding
    val hints = mutableMapOf<DecodeHintType, Any>()
    hints[DecodeHintType.TRY_HARDER] = true
    hints[DecodeHintType.POSSIBLE_FORMATS] = listOf(BarcodeFormat.QR_CODE)

    // Try to decode
    val reader = QRCodeReader()

    return try {
      val result = reader.decode(binaryBitmap, hints)

      // Extract EC level from metadata
      val metadata = result.resultMetadata
      var ecLevel: String? = null

      if (metadata != null) {
        val ecLevelObj = metadata[ResultMetadataType.ERROR_CORRECTION_LEVEL]
        if (ecLevelObj != null) {
          ecLevel = ecLevelObj.toString()
        }
      }

      mapOf(
        "success" to true,
        "data" to result.text,
        "ecLevel" to ecLevel
      )
    } catch (e: NotFoundException) {
      // Try with inverted image
      tryWithInvertedImage(source, hints)
    } catch (e: Exception) {
      mapOf(
        "success" to false,
        "error" to (e.message ?: "Decoding failed")
      )
    } finally {
      bitmap.recycle()
    }
  }

  private fun tryWithInvertedImage(
    source: RGBLuminanceSource,
    hints: Map<DecodeHintType, Any>
  ): Map<String, Any?> {
    return try {
      val invertedSource = source.invert()
      val binaryBitmap = BinaryBitmap(HybridBinarizer(invertedSource))
      val reader = QRCodeReader()
      val result = reader.decode(binaryBitmap, hints)

      val metadata = result.resultMetadata
      var ecLevel: String? = null

      if (metadata != null) {
        val ecLevelObj = metadata[ResultMetadataType.ERROR_CORRECTION_LEVEL]
        if (ecLevelObj != null) {
          ecLevel = ecLevelObj.toString()
        }
      }

      mapOf(
        "success" to true,
        "data" to result.text,
        "ecLevel" to ecLevel
      )
    } catch (e: Exception) {
      mapOf(
        "success" to false,
        "error" to "No QR code found"
      )
    }
  }
}
